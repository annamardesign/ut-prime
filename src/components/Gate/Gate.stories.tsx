import React from 'react';
import type { Meta } from '@storybook/react';

// @ts-ignore: md file and not a module
import page from './README.mdx';
import Gate from './index';

const meta: Meta = {
    title: 'Internal/Gate',
    component: Gate,
    parameters: {docs: {page}},
    args: {
        state: {
            login: {
                get: item => ({
                    cookieChecked: true,
                    authenticated: true,
                    result: {
                        getIn: () => {}
                    }
                }[item])
            }
        }
    }
};
export default meta;

export const Basic: React.FC<{}> = () =>
    <Gate><div className='p-component'>Gate body</div></Gate>;
