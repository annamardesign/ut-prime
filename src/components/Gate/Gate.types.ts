import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';

export interface Props extends React.HTMLAttributes<HTMLDivElement> {
    cookieChecked: boolean;
    gateLoaded: boolean;
    isLogout: boolean;
    authenticated: boolean;
    result: any;
    logout: () => void;
    cookieCheck: ({appId: string}) => void;
    fetchTranslations: (params: {languageId: number, dictName: string[]}) => Promise<any>;
}

const styles = createStyles({
    gate: {
        height: '100%'
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
