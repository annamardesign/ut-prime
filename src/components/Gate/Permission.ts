import PropTypes from 'prop-types';
import { Component } from 'react';

class Permission extends Component {
    static childContextTypes = {
        translate: PropTypes.func,
        money: PropTypes.func,
        dateFormat: PropTypes.func,
        numberFormat: PropTypes.func,
        checkPermission: PropTypes.func,
        implementationStyle: PropTypes.object
    };

    getChildContext() {
        return {
            translate: text => text,
            dateFormat: text => text,
            numberFormat: text => text,
            checkPermission: action => true,
            money: text => text,
            implementationStyle: {}
        };
    }

    render() {
        return this.props.children;
    }
}

export default Permission;
