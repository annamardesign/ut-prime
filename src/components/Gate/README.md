# Gate

Internal component, no do not use it directly

It is used to check login state and perform initial
fetching of data.
