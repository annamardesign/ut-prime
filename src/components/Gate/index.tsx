import React, { useState, useEffect } from 'react';
import { connect, useSelector } from 'react-redux';
import { useParams, Redirect } from 'react-router-dom';

import { cookieCheck } from '../Login/actions';
import Loader from '../Loader';
import Context from '../Text/context';

import Permission from './Permission';
import { Styled, StyledType } from './Gate.types';

const fetchTranslations = params => ({
    type: 'core.translation.fetch',
    method: 'core.translation.fetch',
    params
});

const Gate: StyledType = ({ classes, children, cookieCheck, fetchTranslations, loginPage }) => {
    const [loaded, setLoaded] = useState(null);
    const [cookieChecked, setCookieChecked] = useState(false);
    const login = useSelector(state => state.login || false);
    const loginAbsolute = loginPage && (loginPage.startsWith('http://') || loginPage.startsWith('https://'));

    async function load() {
        // setPermissions(result.get('permission.get').toJS());
        const language = login?.language?.languageId;
        const translations = await fetchTranslations({
            languageId: language,
            dictName: ['text', 'actionConfirmation']
        });
        const dictionary = translations?.result?.translations?.reduce(
            (prev, {dictionaryKey, translatedValue}) => dictionaryKey === translatedValue ? prev : {...prev, [dictionaryKey]: translatedValue},
            {}
        );
        setLoaded({
            language,
            translate: text => dictionary?.[text] || text
        });
    }

    const {appId} = useParams();

    async function check() {
        await cookieCheck({ appId });
        setCookieChecked(true);
    }

    useEffect(() => {
        if (!cookieChecked && !login) {
            check();
        } else if (!loaded && login) {
            load();
        } else if (loginAbsolute && !login) {
            window.location.href = loginPage;
        } else if (loaded && !login) {
            setLoaded(false);
        }
    }, [cookieChecked, login, loaded]);

    if (!cookieChecked && !login) {
        return <Loader />;
    } else if (login) {
        return (
            <div className={classes.gate}>
                {loaded ? <Context.Provider value={loaded}>
                    <Permission>
                        {children}
                    </Permission>
                </Context.Provider> : <Loader />}
            </div>
        );
    } else if (loginAbsolute) {
        return <Loader open message='Redirecting to the login page...' />;
    } else {
        return <Redirect to={loginPage || '/login'} />;
    }
};

export default connect(
    null,
    { cookieCheck, fetchTranslations }
)(Styled(Gate));
