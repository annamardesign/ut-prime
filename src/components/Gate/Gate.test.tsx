import React from 'react';

import { render } from '../test';
import Gate from './index';

describe('<Gate />', () => {
    it('render equals snapshot', async() => {
        const { getByTestId, findByText } = render(
            <Gate>ready</Gate>, {
                login: {},
                loader: {
                    open: true,
                    message: 'Loading...'
                }
            }
        );
        expect(getByTestId('ut-front-test')).toMatchSnapshot('Loading');
        await findByText('ready');
        expect(getByTestId('ut-front-test')).toMatchSnapshot('Ready');
    });
});
