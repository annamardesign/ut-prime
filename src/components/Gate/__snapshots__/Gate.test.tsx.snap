// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gate /> render equals snapshot: Loading 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="Gate-gate-1"
  >
    <div
      class="Loader-loaderContainer-3"
    >
      <div
        class="Loader-overlay-4"
      />
      <div
        class="Loader-loader-5"
      />
      <div
        class="Loader-message-6"
      >
        <span>
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Gate /> render equals snapshot: Ready 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="Gate-gate-1"
  >
    ready
  </div>
</div>
`;
